In [1]:
import csv
import xlrd
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# CPS Historical Migration/Geographic Mobility Tables
https://www.census.gov/data/tables/time-series/demo/geographic-mobility/historic.html
- Note: 
    - As of November 2018, these historical tables include corrections to errors that were discovered. 
    - Please replace any previously downloaded historical tables.
- Action:
    - load in tables
    - brief pre-exploration overview analysis (6 total)

In [2]:
# Table A-1. Annual Geographical Mobility Rates, By Type of Movement: 2018
annual_geographical_mobility_rates_type_movement = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-1.xls' )

# Table A-2. Annual Inmigration, Outmigration, Net Migration, and Movers from Abroad for Regions: 2018
annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-2.xls' )

# Table A-3. Inmigration, Outmigration, and Net Migration by Metropolitan Status: 2018
inmigration_outmigration__net_migration_metropolitan_status = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-3.xls' )

# Table A-4. Geographical Mobility by Tenure: 2018
geographical_mobility_by_tenure = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-4.xls' )

# Table A-5. Reason for Move (Specific Categories): 2018
reason_for_move_specific_categories = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-5.xls' )

# Table A-6. Distance of Intercounty Move: 2018
distance_of_intercounty_move = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-6.xls' )

In [3]:
'''annual_geographical_mobility_rates_type_movement'''

print( annual_geographical_mobility_rates_type_movement.info() )
# annual_geographical_mobility_rates_type_movement.sample( 3 ) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 186 entries, 0 to 185
Data columns (total 10 columns):
Table with row headers in column A and column headers in rows 5 through 7.    154 non-null object
Unnamed: 1                                                                    137 non-null object
Unnamed: 2                                                                    137 non-null object
Unnamed: 3                                                                    137 non-null object
Unnamed: 4                                                                    138 non-null object
Unnamed: 5                                                                    137 non-null object
Unnamed: 6                                                                    138 non-null object
Unnamed: 7                                                                    137 non-null object
Unnamed: 8                                                                    137 non-null object
Unnamed: 9   

***annual_geographical_mobility_rates_type_movement***
- notes:
    - seems to cover back to 1947
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom
    - varying length of information
        - count (column(s))
            - 154 (0) , 138 (4, 6) , 137 (1, 2, 3, 5, 7, 8, 9)
        - information length is most likely <= 137

In [4]:
'''annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions'''

# print( annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions.info() )
annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions.sample( 3 )

Table with row headers in column A and column headers in row 5.  \
27                                         .Inmigrants                
8                              .Net internal migration                
284                                .Movers from abroad                

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4  
27         403        463       1010        562  
8        -352*        -38       512*       -122  
284        229        134        401        324

***annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions***
- notes:
    - seems to cover back to 1980
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom

In [5]:
'''inmigration_outmigration__net_migration_metropolitan_status'''

# print( inmigration_outmigration__net_migration_metropolitan_status.info() )
inmigration_outmigration__net_migration_metropolitan_status.sample( 3 )

Table with row headers in column A and column headers in row 5.  \
43                         2010-2011 (2010 controls)/3                
105                             .Nonmetropolitan areas                
119                                          1998-1999                

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5  
43         NaN        NaN        NaN        NaN        NaN  
105       1825       2178      -353*        127      -225*  
119        NaN        NaN        NaN        NaN        NaN

***inmigration_outmigration__net_migration_metropolitan_status***
- notes:
    - seems to cover back to 1985
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom

In [6]:
'''geographical_mobility_by_tenure'''

# print( geographical_mobility_by_tenure.info() )
geographical_mobility_by_tenure.sample( 3 )

Table with row headers in column A and column headers in rows 5 through 7.  Numbers start in row 10, percentages start in row 184, and frequency distributions start in row 358.  \
300                         ..In renter-occupied units                                                                                                                                 
317                                        1994-1995/6                                                                                                                                 
20                           ..In owner-occupied units                                                                                                                                 

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
300        100       66.8       33.2       31.9       19.7       12.2   
317        NaN        NaN        NaN        NaN        NaN        NaN   
20      206767     196431      10336      10135       6278       3857   

    Unnamed: 7 Unnamed: 8 Unnamed: 9  
300        6.1        6.1        1.3  
317        NaN        NaN        NaN  
20        2608       1249        201

***geographical_mobility_by_tenure***
- notes:
    - seems to cover back to 1987
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom

In [7]:
'''reason_for_move_specific_categories'''

# print( reason_for_move_specific_categories.info() )
reason_for_move_specific_categories.sample( 3 )

Table with row headers in columns A and K, and column headers in rows 5 and 6. Numbers start in row 9, percentages start in row 36.  \
3                                     Mobility Period                                                                                    
2                   (leading dots indicate sub-parts)                                                                                    
45                        2009-2010 (2000 controls)/4                                                                                    

                       Unnamed: 1              Unnamed: 2 Unnamed: 3  \
3   Total Movers 1 year and over   Family-related reasons        NaN   
2                             NaN                     NaN        NaN   
45                            100                  7.3415    11.2174   

   Unnamed: 4                  Unnamed: 5 Unnamed: 6 Unnamed: 7 Unnamed: 8  \
3         NaN  Employment-related reasons        NaN        NaN        NaN   
2         NaN                         NaN        NaN        NaN        NaN   
45    11.7448                     7.79968    2.58657    4.23548   0.527437   

   Unnamed: 9  ... Unnamed: 12 Unnamed: 13 Unnamed: 14 Unnamed: 15  \
3         NaN  ...         NaN         NaN         NaN         NaN   
2         NaN  ...         NaN         NaN         NaN         NaN   
45    1.29995  ...     15.4502     4.07565     10.8338           -   

   Unnamed: 16    Unnamed: 17 Unnamed: 18 Unnamed: 19 Unnamed: 20 Unnamed: 21  
3          NaN  Other reasons         NaN         NaN         NaN         NaN  
2          NaN            NaN         NaN         NaN         NaN         NaN  
45     8.72403         2.7464    0.631327     1.50506    0.279702     4.38466  

[3 rows x 22 columns]

***reason_for_move_specific_categories***
- notes:
    - seems to cover back to 1998
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom

In [8]:
# distance_of_intercounty_move
# print( distance_of_intercounty_move.info() )
distance_of_intercounty_move.sample( 3 )

Table with row headers in column A and column headers in row 5. Numbers start in row 8, percentages start in row 32.  \
23                                                NaN                                                                     
38                        2010-2011 (2010 controls)/2                                                                     
43                                          2008-2009                                                                     

   Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5  
23        NaN        NaN        NaN        NaN        NaN  
38        100       38.6       21.3       13.9       26.2  
43        100    39.9402    21.7691    14.2106    24.0801

***distance_of_intercounty_move***
- notes:
    - seems to cover back to 2002
    - needs to be cleaned , pd does not naturally detect formatting 
        - top and bottom

## Aggregate Initial Notes
- ***Dates***
    - first recording year varies by up to 55 years 

| first recording year | dataset |
|- - - -|- - - -|
| 1947 | annual_geographical_mobility_rates_type_movement |
|- - - -|- - - -|
| 1980 | annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions |
|- - - -|- - - -|
| 1985 | inmigration_outmigration__net_migration_metropolitan_status |
|- - - -|- - - -|
| 1987 | inmigration_outmigration__net_migration_metropolitan_status |
|- - - -|- - - -|
| 1998 | reason_for_move_specific_categories |
|- - - -|- - - -|
| 2002 | distance_of_intercounty_move |

- ***format***
    - first rows of each file seem to serve as information , actual tables start near end of .head()
    - ending rows of each file seem to serve as annotation and disclaimer
- ***actions***:
    - walk though one table at a time
    - let's start with dropping NaN clustered rows

In [9]:
# copy dataframe , generalize name
df = annual_geographical_mobility_rates_type_movement.copy()
# drop columns without any non-NaN values
df = df.dropna(thresh= ( 1 ) )
# return the dataframe to see results
# df
df.sample(7)

Table with row headers in column A and column headers in rows 5 through 7.  \
185  Source:  U.S. Census Bureau, Current Populatio...                           
44                                           1991-1992                           
64                                           1967-1968                           
4                                                  NaN                           
92                                           2017-2018                           
31                  2000-2001 (SCHIP, 2000 controls)/5                           
158                                          1959-1960                           

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4   Unnamed: 5  \
185        NaN        NaN        NaN        NaN          NaN   
44      247380     204580      42800      41545        26587   
64      194621     156735      37886      36603        22960   
4          NaN        NaN        NaN      Total  Same county   
92         100       89.9       10.1        9.8          6.2   
31      275611     235726      39885      38133        22774   
158        100       80.1       19.9       19.4         12.9   

           Unnamed: 6 Unnamed: 7 Unnamed: 8 Unnamed: 9  
185               NaN        NaN        NaN        NaN  
44              14957       7853       7105       1255  
64              13643       6607       7035       1283  
4    Different county        NaN        NaN        NaN  
92                3.6          2        1.5        0.4  
31              15360       7778       7582       1752  
158               6.4        3.3        3.2        0.5

- ***notes:***
    - 156 rows remain
    - some column names in row 3 , some column names in row 4 , some column names in row 5
        row 3 = 6 column names
            columns = 0 , 1 , 2 , 3 , 4 , 9
        row 4 = 2 column names
            columns = 5 , 6
        row 5 = 2 column names
            columns = 7 , 8
    - row 4 column 4 and row 5 column 6 contain 'Total'
        not obvious what this is
        but it is not the column name in either case
        in both cases , the column name is one row higher in the same column
    - rows 173 - 185 contain only 1 NaN value
- ***actions:***
    - drop all rows without 2 or more non-NaN values
    - merge rows 3 , 4 , and 5 
        taking out 'Total'

In [10]:
# rename dataframe
df = annual_geographical_mobility_rates_type_movement.copy()

# drop columns with less than 2 non-NaN values
df = df.dropna( thresh= ( 2 ) )

# collect column names as pandas series
col_names_0to4_9 = df.iloc[ 0 ]
col_names_5to6 = df.iloc[ 1 ]
col_names7to8 = df.iloc[ 2 ]

# initialize list for column names
column_names = []

# fill list with first row of column names , no exceptions
for column_name in col_names_0to4_9:    
    column_names.append( column_name )

# set index count to -1 so column 0 == column 0
_ = -1

# replace items 5 and 6 in column_names with column names 5 and 6
for column_name in col_names_5to6:
    _ += 1
    # column_name is string (nan instances = float) and name is not 'Total'   
    if isinstance( column_name , str ) and column_name != 'Total':
        # column_name is aligned to replace a nan         
        if str( column_names[ _ ] ) == 'nan':
            # replace item in column_names with column_name             
            column_names[ _ ] = column_name

# set index count to -1 so column 0 == column 0
_ = -1

# replace items 7 and 8 in column_names with column names 7 and 8
for column_name in col_names7to8:
    _ += 1
    # column_name is string (nan instances = float) and name is not 'Total'   
    if isinstance( column_name , str ) and column_name != 'Total':
        # column_name is aligned to replace a nan         
        if str( column_names[ _ ] ) == 'nan':
            # replace item in column_names with column_name             
            column_names[ _ ] = column_name

# (re)set column names
df.columns = column_names

# drop rows used to extract column names
df = df[3:]

# reset dataframe index
df = df.reset_index( drop=True ) 

# let's see
df.head(10)

Mobility period Total, 1 year old and over  \
0                          2017-2018                     319319   
1                          2016-2017                     316550   
2                          2015-2016                     314992   
3                          2014-2015                     312295   
4  2013-2014 (98,000 address file)/1                     309601   
5                          2012-2013                     307243   
6                          2011-2012                     304924   
7        2010-2011 (2010 controls)/2                     302640   
8        2010-2011 (2000 controls)/3                     302005   
9        2009-2010 (2010 controls)/2                     300419   

  Same residence (non-movers) Total movers  \
0                      286967        32352   
1                      281647        34902   
2                      279854        35138   
3                      275971        36324   
4                      273920        35681   
5                      271325        35918   
6                      268436        36488   
7                      267602        35038   
8                      266930        35075   
9                      262975        37445   

  Different residence in the United States Same county Different county  \
0                                    31186       19852            11334   
1                                    33647       21614            12033   
2                                    33857       21588            12269   
3                                    34651       22905            11746   
4                                    34548       23436            11112   
5                                    34882       23150            11731   
6                                    35334       23493            11842   
7                                    33953       23330            10623   
8                                    34016       23325            10691   
9                                    36459       25910            10549   

  Same state Different state Movers from abroad  
0       6496            4838               1165  
1       6668            5366               1255  
2       7501            4768               1281  
3       6653            5093               1673  
4       6446            4666               1133  
5       6961            4770               1036  
6       6782            5059               1154  
7       5868            4756               1084  
8       5912            4779               1058  
9       6227            4323                985

- ***as of now***:
    - removed rows with NaN values
    - extracted, sorted, and set column names
- ***next (action)***:
    - we need to clean up the 'Mobility period' column
        - samples from current:
            - 2017-2018
            - 2013-2014 (98,000 address file)/1
            - 2011-2012
            - 2010-2011 (2010 controls)/2
            - 2000-2001 (SCHIP, 2000 controls)/5
            - 1970-1971/8
        - what we want:
            - 2018 << 2017-2018
            - 2014 << 2013-2014 (98,000 address file)/1
                requires further investigation
            - 2012 << 2011-2012
            - 2011 << 2010-2011 (2010 controls)/2
                requires further investigation
            - 2001 << 2000-2001 (SCHIP, 2000 controls)/5
                requires further investigation
            - 1971 << 1970-1971/8

In [195]:
# examine data in 'Mobility period'
# _ = 0
# for instance in df[ 'Mobility period' ]:
#     # print( _ , instance )
#     _ += 1
    
# selected outputs (to avoid long out)
'''
# non-uniform , non-repeated
    3 2014-2015
    4 2013-2014 (98,000 address file)/1
    5 2012-2013
    
# repeats w/ controls , no uniform representation
    6 2011-2012
    7 2010-2011 (2010 controls)/2
    8 2010-2011 (2000 controls)/3
    9 2009-2010 (2010 controls)/2
    10 2009-2010 (2000 controls)/3
    11 2008-2009

    18 2001-2002
    19 2000-2001 (2000 controls)/3
    20 2000-2001 (SCHIP, 2000 controls)/5
    21 2000-2001 (1990 controls)/6
    22 1999-2000
    
    28 1993-1994
    29 1992-1993 (1990 controls)/6
    30 1992-1993 (1980 controls)/7
    31 1991-1992
    
# lack of info from 70s 
    41 1981-1982
    42 1980-1981
    43 1975-1976/8
    44 1970-1971/8
    45 1969-1970
    46 1968-1969
    47 1967-1968
    
# in chronological order until row 68 (then repeats)
    66 1948-1949
    67 1947-1948
    68 2017-2018
    69 2016-2017
'''
# _ = 68
# make sure rows 0-67 are the same as 68-135
# for row in df[ 'Mobility period' ][ 0 : 68 ]:
#     _ += 1
#     if row != df[ 'Mobility period' ][ _ ]:
#         print(row , df[ 'Mobility period' ][ _ ])
# _ = 0
# for  in df[ 'Mobility period' ]:
#     print(_,row)
#     _+=1
copy_mobility_period = df[ 'Mobility period' ].copy()
for _ in range( len( copy_mobility_period ) ):
    print( copy_mobility_period.count( ( copy_mobility_period[ _ ] ) ) )


AttributeError: 'RangeIndex' object has no attribute 'levels'

In [12]:
# determine type of data in 'Mobility period' 
def test_type( dataframe , test_type ):
    '''
    input) dataframe w/ specified column 
        - e.g. df['specified column']
    input) type wanting to test
        - e.g. str 
    
    output) count of instances which are not type == test_type
    '''
    non_test_type = 0
    for _ in dataframe:
        if isinstance( _ , test_type ) == False:
            non_test_type += 1
    return non_test_type


test_000 = test_type( dataframe=df[ 'Mobility period' ] , test_type=str )
test_000

0

- ***as test_000 returns 0***:
    - the type of all datapoints in 'Mobility period' is str (string)
    - we will not need to change any points to have uniformity   
- ***next up (action)***:
    - clear low-hanging fruit
    - change the simple ranges to int for set year
        - e.g. change "2017-2018" to "2018"

In [13]:
# identify changes which can be implimented without further investigation
mobility_period = df[ 'Mobility period' ]
simple_year_ranges_count_ = 0
confirmed_simple_year_ranges_count_ = 0

# count how many simple-fix years we have
for datapoint in mobility_period:
    # len('2012-2013') == 9 , check datapoint is formatted as expected     
    if len( datapoint ) == 9 and len( datapoint.split( '-' ) ) == 2:
        # pretty sure we know what's going on but will investigate further         
        simple_year_ranges_count_ += 1
        # first 4 characters of string , can be converted to int
        first_year = int( datapoint[ :4 ] )
        # last 4 characters of string , can be converted to int
        second_year = int( datapoint[ -4: ] )
        # first_year and last_year are post-1940 (because expected), and year range is 1 year 
        if first_year and second_year > 1940 and second_year - first_year == 1:
            # the datapoint is within our risk threshold
            confirmed_simple_year_ranges_count_ += 1

# if everything we thought was what we think it is 
if simple_year_ranges_count_ == confirmed_simple_year_ranges_count_:
    print(f'total: {confirmed_simple_year_ranges_count_}\nratio: {confirmed_simple_year_ranges_count_/len(mobility_period)}')

total: 110
ratio: 0.8088235294117647


- ***notes***:
    - know count of lowest fruit to be 110 ( > 80.88% , not bad)
    - just confirmed status of 110 points to be (at least) similar enough to extract years
    - knowing simple_year_ranges_count_ allows the conformation of outcome via count compairson 
- ***action***:
    - convert those 110 year ranges to years represented by the 2nd year in each range

In [14]:
# set empty
less_low_mobility_periods = []
# so row 0 == row 0
_ = -1

# little bit of ctrl + c , little bit of ctrl + v
for datapoint in mobility_period:
    _ += 1
    # len('2012-2013') == 9 , check datapoint is formatted as expected     
    if len( datapoint ) == 9 and len( datapoint.split( '-' ) ) == 2:
        # replace datapoint w/ int of last 4 characters of string 
        mobility_period[ _ ] = int( datapoint[ -4: ] )
    else:
        # keep what's already there
        less_low_mobility_periods.append( datapoint )

mobility_period.sample(7)

123                           1960
9      2009-2010 (2010 controls)/2
59                            1956
112                    1970-1971/8
50                            1965
97     1992-1993 (1990 controls)/6
64                            1951
Name: Mobility period, dtype: object

- ***notes***:
    - we have converted the low hanging fruit
- ***actions***:
    - now on to exploring the other 26 rows

In [172]:
# print them out , let's see what we can see
_ = 0
for row in df['Mobility period']:
    _ += 1    
    if not isinstance( row , int ):
        # _ check for consistent spacing
        if _ < 10:
            print( str(_) + '  |  ' + row )
        elif _ < 100:
            print( str(_) + ' |  ' + row )
        else:
            print( str(_) + '|  ' + row )

5  |  2013-2014 (98,000 address file)/1
8  |  2010-2011 (2010 controls)/2
9  |  2010-2011 (2000 controls)/3
10 |  2009-2010 (2010 controls)/2
11 |  2009-2010 (2000 controls)/3
16 |  2004-2005/4
20 |  2000-2001 (2000 controls)/3
21 |  2000-2001 (SCHIP, 2000 controls)/5
22 |  2000-2001 (1990 controls)/6
30 |  1992-1993 (1990 controls)/6
31 |  1992-1993 (1980 controls)/7
44 |  1975-1976/8
45 |  1970-1971/8
73 |  2013-2014 (98,000 address file)/1
76 |  2010-2011 (2010 controls)/2
77 |  2010-2011 (2000 controls)/3
78 |  2009-2010 (2010 controls)/2
79 |  2009-2010 (2000 controls)/3
84 |  2004-2005/4
88 |  2000-2001 (2000 controls)/3
89 |  2000-2001 (SCHIP, 2000 controls)/5
90 |  2000-2001 (1990 controls)/6
98 |  1992-1993 (1990 controls)/6
99 |  1992-1993 (1980 controls)/7
112|  1975-1976/8
113|  1970-1971/8


- ***notes***
    - 6 rows formatted almost like majority (new lowest)
        - all have either /4 or /8 at the end of otherwise standard string
            - e.g. standard 
                - 2004-2005
            - observation rows 
                - 16 |  2004-2005/4
                - 44 |  1975-1976/8
                - 45 |  1970-1971/8
                - 84 |  2004-2005/4
                - 112|  1975-1976/8
                - 113|  1970-1971/8
        - remarks before reading documentation
            - identical instances
                - rows 16 and 84
                - rows 44 and 112
                - rows 45 and 113
            - rows are not ordered by year
                - unless years reoccour/loop through table
                - or otherwise
    - 16 rows are in the next new lowest format
        - all contain ({year} controls)/{num} with year and num in consistent pairs
            - (year / num) pairs = # occurrences
                - (2010 / 2) = 4
                - (2000 / 3) = 6
                - (1990 / 6) = 4
                - (1980 / 7) = 2
        - they occour in pattern twice:
            - 4 uninterrupted
            - 1
            - 3 uninterrupted
        - sample rows
            - 8  |  2010-2011 (2010 controls)/2
            - 9  |  2010-2011 (2000 controls)/3
            - 10 |  2009-2010 (2010 controls)/2
            - 11 |  2009-2010 (2000 controls)/3
        - unknown analysis
            - seems to be game settings for different censuses 
            - difficult to connect the /2,/3 and /6,/7 relationship
                - if was dependent on decade order , possibles would be straight ordered 
                    - e.g. /2, /3, /4, /5 or /1, /3, /5, /7
                    
    - final 4 rows are longer and are 2 duplicate pairs
        - default format + ending either (98,000 address file)/1 or (SCHIP, 2000 controls)/5
            - longest rows
                - 5  |  2013-2014 (98,000 address file)/1
                - 21 |  2000-2001 (SCHIP, 2000 controls)/5
                - 73 |  2013-2014 (98,000 address file)/1
                - 89 |  2000-2001 (SCHIP, 2000 controls)/5
        - pre document remarks
            - first notice of /1 
            - return use of /5 
                - do other /n have additional possibile (paramaters)
    - blind analysis
        - year range
                - min: 1970-1971 (1971 in complete df)
                - max: 2013-2014 (2014 in complete df)
            - less range than full representation of column
                - min: 1948 (1947-1948 in origional df)
                - max: 2018 (2017-2018 in origional df)
            - why?
                - recent changes to census
                    - collection
                    - questions
                    - coverage
                - recent increase in census related surveys
                    - housing surveys
                    - business surveys
                        - we will be exploring these after fitting gen_000 model
                        - these are also how this data is presented for non-decade years
        - lot of possible meanings for non-conforming objects
                - politics..
                - changing standards
                - technological evolution
                - etc...
                    - storage , collection , methods , etc... 
            - full-court lobs
                - something to do with immigrants
                - evolution of process 
                - expansion of non-physical opportunities 
        - there's got to be a better way to extract dataframes from excel files
            - likely convertable, but annotations likely direct either universe to same result 
- ***actions***
    - go find out what these non-standard occurrences mean

In [16]:
# make another copy of origional dataframe (this is why we copy)
info_df = annual_geographical_mobility_rates_type_movement.copy()

# function to make NaN dataframe
row_nans = lambda df: df[ df.isnull().any( axis=1 ) ]

# identify the rows with information (basically, keep the NaNs this time)
nan_df = row_nans( info_df )

# peak some rows with NaN values 
nan_df.head()

Table with row headers in column A and column headers in rows 5 through 7.  \
0  Table A-1. Annual Geographical Mobility Rates,...                           
1                            (Numbers in thousands.)                           
2                                                NaN                           
3                                    Mobility period                           
4                                                NaN                           

                   Unnamed: 1                   Unnamed: 2    Unnamed: 3  \
0                         NaN                          NaN           NaN   
1                         NaN                          NaN           NaN   
2                         NaN                          NaN           NaN   
3  Total, 1 year old and over  Same residence (non-movers)  Total movers   
4                         NaN                          NaN           NaN   

                                 Unnamed: 4   Unnamed: 5        Unnamed: 6  \
0                                       NaN          NaN               NaN   
1                                       NaN          NaN               NaN   
2                                       NaN          NaN               NaN   
3  Different residence in the United States          NaN               NaN   
4                                     Total  Same county  Different county   

  Unnamed: 7 Unnamed: 8          Unnamed: 9  
0        NaN        NaN                 NaN  
1        NaN        NaN                 NaN  
2        NaN        NaN                 NaN  
3        NaN        NaN  Movers from abroad  
4        NaN        NaN                 NaN

In [17]:
# noted column 0 name 
column_names = []

# make it slightly educational
for _ in nan_df[ :0 ]:
    # each _ is a string
    if isinstance( _ , str ):
        # add the string to column_names
        column_names.append( _ )
        
# this is the education
print( f'len(nan_df[:0])=={ len( nan_df[ :0 ] ) }\ntype(nan_df[:0])=={ type( nan_df[ :0 ] ) }\n'
      # The economy, stupid. 
       f'len(column_names)=={ len( column_names ) }\ntype(column_names={ type( column_names ) }' )

len(nan_df[:0])==0
type(nan_df[:0])==<class 'pandas.core.frame.DataFrame'>
len(column_names)==10
type(column_names=<class 'list'>


In [18]:
# we can find row headers in column A
if column_names[0] != 'Table with row headers in column A and column headers in rows 5 through 7.':
    raise AssertionError('')

# function to return non-NaN rows 
non_nans = lambda df: df.dropna()

# identify non-NaN rows from column A (the first column)
column_a = non_nans( nan_df[ 'Table with row headers in column A and column headers in rows 5 through 7.' ] )

# let's see
column_a

0      Table A-1. Annual Geographical Mobility Rates,...
1                                (Numbers in thousands.)
3                                        Mobility period
6                                                 NUMBER
90                                               PERCENT
173                                           Footnotes:
174    1/ The 2014 CPS ASEC included redesigned quest...
175    2/ Population controls consistent with 2010 Ce...
176    3/ Population controls consistent with 2000 Ce...
177    4/ Caution should be used when comparing numbe...
178    5/ SCHIP sample: poulation controls consistent...
179    6/ Population controls consistent with 1990 Ce...
180    7/ Population controls consistent with 1980 Ce...
181    8/ The 1-year geographic mobility question was...
182    For more information on the transition from Ce...
183    Adjustments to Household Survey Population Est...
184       Population Estimates in January 2012 webpage\n
185    Source:  U.S. Census Bur

In [19]:
# identify information of interest
indicators_of_interest = [ '1/' , '2/' , '3/' , '4/' , '5/' , '6/' , '7/' , '8/' ]

# make copy
copy_column_a = column_a.copy()

# return rows from copy_column_a which contain an indicator from indicators_of_interest
rows_of_interest = [ row for indicator in indicators_of_interest for row in copy_column_a if indicator in row ]

# examine the lowest (remember: /4 and /8)
for row in rows_of_interest:
    if '4/' in row or '8/' in row:
        print( row + '\n' )

4/ Caution should be used when comparing numbers/rates of movers within the same county and from a different county between the 1999-2000 to 2004-2005 period with other periods. A change in processing resulted in a shift in the distribution of movers for these years. For additional information, see the "Impact of Processing on CPS Interstate Migration Rates" note < http://www.census.gov/population/www/socdemo/CPSnote.pdf >.   

8/ The 1-year geographic mobility question was not asked between 1972 through 1975 and 1977 to 1980. The first half of the decade (1971-1975) asked about migration since 1970 and the second half (1976-1980) since 1975. Therefore, only 1970-1971 and 1975-1976 ask a comparable 1-year question.    



- ***notes***:
    - /4 is fairly easy to account for
        - basically they flipped the method of recording 
            - https://www.census.gov/population/www/socdemo/CPSnote.pdf 
            - from
                - group > geography
            - to 
                - geography > group
        - "Users of interstate migration data should be cautious of comparing rates from the 1999-2000 to 2004-5 period with other periods."
    - /8 is dissapointing due to prelim scanning of San Francisco 
        - will need to be remembered if one year are factored
        - also indicates 5-year question could be used 
            - would need to ensure other 5-year periods are/get sorted 
            
- ***actions***:
    - compare with neighbors to determine optimal implementation 
        - hedge repeats

In [86]:
# reset
sims = df['Mobility period']
_ = -1

for row in sims:
    # update row
    _ += 1
    # id indicators
    if '/4'  in str(row) or '/8' in str(row): 
        # ensure next is logical
        if _ < 2:
            raise Exception( 'f' )
        # print 2 before through 2 after
        print ( '\n\n\n\n' , sims[ _-2 : _+3 ] )





 13           2007
14           2006
15    2004-2005/4
16           2004
17           2003
Name: Mobility period, dtype: object




 41           1982
42           1981
43    1975-1976/8
44    1970-1971/8
45           1970
Name: Mobility period, dtype: object




 42           1981
43    1975-1976/8
44    1970-1971/8
45           1970
46           1969
Name: Mobility period, dtype: object




 81           2007
82           2006
83    2004-2005/4
84           2004
85           2003
Name: Mobility period, dtype: object




 109           1982
110           1981
111    1975-1976/8
112    1970-1971/8
113           1970
Name: Mobility period, dtype: object




 110           1981
111    1975-1976/8
112    1970-1971/8
113           1970
114           1969
Name: Mobility period, dtype: object


- ***notes***:
    - no /4 and /8 instances have duplicate year range  
    - we are reliant on both /4 and /8 instances (all)
    
- ***actions***:
    - continue examining other non-standard rows
    - add /4 and /8 to list for default conversion 
        - default conversion = note in new column & convert as if standard

In [79]:
# make copy of rows of interest
copy_rows_of_interest = rows_of_interest.copy()
# used: /4, /8 ; unused: 1/, /2, /3, /5, /6, /7

# examine the next lowest (remember: /2 and /3 ; largest set, size=16)
for row in copy_rows_of_interest:
    if '2/' in row or '3/' in row:
        print( row + '\n' )

2/ Population controls consistent with 2010 Census.

3/ Population controls consistent with 2000 Census.



- ***notes***:
    - requires conformation of duplicates but ideally move forward with 2010 controls 
        - values most recent measures
        - return for 2000 controls for comps if time
    - https://www.census.gov/content/dam/Census/programs-surveys/acs/guidance/2010_Change_Population_Controls.pdf 
- ***actions***:
    - confirm duplicates 

In [171]:
# reset
sims_001 = df['Mobility period']

# for each instance in sims_001
for _ in range( len( sims_001 ) ):
    # if instance is string
    if isinstance( sims_001[ _ ] , str ):
        # and first 11 characters in that string are same as first 11 characters in str(next row)
            # e.g. 2010-2011 ( == 2010-2011 (
        if sims_001[ _ ][ :11 ] == str( sims_001[ _+1 ])[ :11 ]:
            # output the first row, then the second
            print( sims[ _ ], '\n', sims[ _+1 ], '\n' )

2010-2011 (2010 controls)/2 
 2010-2011 (2000 controls)/3 

2009-2010 (2010 controls)/2 
 2009-2010 (2000 controls)/3 

2000-2001 (2000 controls)/3 
 2000-2001 (SCHIP, 2000 controls)/5 

2000-2001 (SCHIP, 2000 controls)/5 
 2000-2001 (1990 controls)/6 

1992-1993 (1990 controls)/6 
 1992-1993 (1980 controls)/7 

2010-2011 (2010 controls)/2 
 2010-2011 (2000 controls)/3 

2009-2010 (2010 controls)/2 
 2009-2010 (2000 controls)/3 

2000-2001 (2000 controls)/3 
 2000-2001 (SCHIP, 2000 controls)/5 

2000-2001 (SCHIP, 2000 controls)/5 
 2000-2001 (1990 controls)/6 

1992-1993 (1990 controls)/6 
 1992-1993 (1980 controls)/7 



- ***notes***:
    - duplicate confirmed:
        - 2010-1011
        - 2009-2010
        - 1992-1993        
        - each duplicate repeated in line with table repeat...
    - duplicate *NOT* confirmed:
        - /2 or /3 dependent 
            - 2000-2001
                - 2000-2001 (2000 controls)/3 
                - 2000-2001 (SCHIP, 2000 controls)/5 
            - duplicate repeated in line with table repeat...
        - *NOT* /2 or /3 dependent
            - 2000-2001
                - 2000-2001 (SCHIP, 2000 controls)/5 
                - 2000-2001 (1990 controls)/6 
            - 1992-1993
                - 1992-1993 (1990 controls)/6 
                - 1992-1993 (1980 controls)/7 
            - each duplicate repeated in line with table repeat...
    - some of these contain /5
        - *but NONE* contain /1
- ***actions***:
    - re-check for duplicate year ranges between other /# 
        - always good practice
        - could lead to reconsideration of one(+) /#

In [193]:
# reset
sims_002 = df[ 'Mobility period' ]
_ = 0
# to remember what we've seen
storage = []

# let's see what we can see
for row in sims_002:
    # keep track of the row
    _ += 1 

    # if instance (row) is not int
    if not isinstance( row , int ):
        
        # add 2nd year in year range to storage
        storage.append( row[ 5:9 ] )
        
        # how many times have we seen that year
        s_c_ = str( storage.count( row[ 5:9 ] ) )
        
        # we only care if the year has been seen at least twice (inclusive) as all rows are seen twice
        if s_c_ >= 2:
            # print row number, 2nd year in year range, times 2nd year has been seen, entire row
            # account for row number when formatting
            if _ < 10:
                print( str( _ ) + '  |  ' + row[ 5:9 ] + '  |  ' + s_c_ + '  |  ' + row )
            elif _ < 100:
                print( str( _ ) + ' |  ' + row[ 5:9 ] + '  |  ' + s_c_ + '  |  ' + row )
            else:
                print( str( _ ) + '|  ' + row[ 5:9 ] + '  |  ' + s_c_ + '  |  ' + row )


5  |  2014  |  1  |  2013-2014 (98,000 address file)/1
8  |  2011  |  1  |  2010-2011 (2010 controls)/2
9  |  2011  |  2  |  2010-2011 (2000 controls)/3
10 |  2010  |  1  |  2009-2010 (2010 controls)/2
11 |  2010  |  2  |  2009-2010 (2000 controls)/3
16 |  2005  |  1  |  2004-2005/4
20 |  2001  |  1  |  2000-2001 (2000 controls)/3
21 |  2001  |  2  |  2000-2001 (SCHIP, 2000 controls)/5
22 |  2001  |  3  |  2000-2001 (1990 controls)/6
30 |  1993  |  1  |  1992-1993 (1990 controls)/6
31 |  1993  |  2  |  1992-1993 (1980 controls)/7
44 |  1976  |  1  |  1975-1976/8
45 |  1971  |  1  |  1970-1971/8
73 |  2014  |  2  |  2013-2014 (98,000 address file)/1
76 |  2011  |  3  |  2010-2011 (2010 controls)/2
77 |  2011  |  4  |  2010-2011 (2000 controls)/3
78 |  2010  |  3  |  2009-2010 (2010 controls)/2
79 |  2010  |  4  |  2009-2010 (2000 controls)/3
84 |  2005  |  2  |  2004-2005/4
88 |  2001  |  4  |  2000-2001 (2000 controls)/3
89 |  2001  |  5  |  2000-2001 (SCHIP, 2000 controls)/5
90 |  200